File: LeetCode/0354. Russian Doll Envelopes/solution.cpp
/**
 * 84 / 84 test cases passed.
 * Runtime: 1100 ms
 * Memory Usage: 16 MB 
 */
class Solution {
public:
    int maxEnvelopes(vector<vector<int>>& envelopes) {
        sort(envelopes.begin(), envelopes.end(), [](const vector<int>& v1, const vector<int>& v2) {
            return v1[0] == v2[0] ? v1[1] > v2[1] : v1[0] < v2[0];
        });
        vector<int> dp(envelopes.size(), 1);
        for (int i = 1; i < envelopes.size(); ++ i) {
            for (int j = 0; j < i; ++ j) {
                if (envelopes[i][1] > envelopes[j][1]) {
                    dp[i] = max(dp[i], dp[j] + 1);
                }
            }
        }
        return *max_element(dp.begin(), dp.end());
    }
};

File: LeetCode/0354. Russian Doll Envelopes/solution.py
"""
56 / 56 test cases passed.
Runtime: 64 ms
Memory Usage: 16.3 MB
"""
import bisect
class Solution:
    def maxEnvelopes(self, envelopes: List[List[int]]) -> int:
        if not len(envelopes):
            return 0

        envelopes.sort(key=lambda x: (x[0], -x[1]))
        record = [envelopes[0][1]]
        for _, h in envelopes[1:]:
            if h > record[-1]:
                record.append(h)
            else:
                idx = bisect.bisect_left(record, h)
                record[idx] = h
        return len(record)
